cpp/include/raft/neighbors/detail/nn_descent.cuh

```diff
@@ -0,0 +1,1447 @@
+  __host__ __device__ bool is_new() const { return id_ >= 0; }
+  __host__ __device__ Index_t& id_with_flag() { return id_; }
+  __host__ __device__ Index_t id() const
+  {
+    if (is_new()) return id_;
+    return -id_ - 1;
+  }
+  __host__ __device__ DistData_t& dist() { return dist_; }
+
+  __host__ __device__ void mark_old()
+  {
+    if (id_ >= 0) id_ = -id_ - 1;
+  }
+
+  __host__ __device__ bool operator<(const ResultItem<Index_t>& other) const
+  {
+    if (dist_ == other.dist_) return id() < other.id();
+    return dist_ < other.dist_;
+  }
+  __host__ __device__ bool operator==(const ResultItem<Index_t>& other) const
+  {
+    return id() == other.id();
+  }
+  __host__ __device__ bool operator>=(const ResultItem<Index_t>& other) const
+  {
+    return !(*this < other);
+  }
+  __host__ __device__ bool operator<=(const ResultItem<Index_t>& other) const
+  {
+    return (*this == other) || (*this < other);
+  }
+  __host__ __device__ bool operator>(const ResultItem<Index_t>& other) const
+  {
+    return !(*this <= other);
+  }
+  __host__ __device__ bool operator!=(const ResultItem<Index_t>& other) const
+  {
+    return !(*this == other);
+  }
+};
+
+constexpr __host__ __device__ size_t div_up(const size_t a, const size_t b)
+{
+  return a / b + (a % b != 0);
+}
+
+constexpr int to_multiple_of_32(int number) { return div_up(number, 32) * 32; }
+
+constexpr int WARP_SIZE          = 32;
+constexpr unsigned int FULL_MASK = 0xffffffff;
+
+template <typename T>
+int get_batch_size(const int it_now, const T nrow, const int batch_size)
+{
+  int it_total = div_up(nrow, batch_size);
+  return (it_now == it_total - 1) ? nrow - it_now * batch_size : batch_size;
+}
+
+// for avoiding bank conflict
+template <typename T>
+constexpr __host__ __device__ __forceinline__ int skew_dim(int ndim)
+{
+  // all "4"s are for alignment
+  if constexpr (std::is_same<T, float>::value) {
+    ndim = div_up(ndim, 4) * 4;
+    return ndim + (ndim % 32 == 0) * 4;
+  }
+}
+
+template <typename T>
+__device__ __forceinline__ ResultItem<T> xor_swap(ResultItem<T> x, int mask, int dir)
+{
+  ResultItem<T> y;
+  y.dist()         = __shfl_xor_sync(FULL_MASK, x.dist(), mask, WARP_SIZE);
+  y.id_with_flag() = __shfl_xor_sync(FULL_MASK, x.id_with_flag(), mask, WARP_SIZE);
+  return x < y == dir ? y : x;
+}
+
+__device__ __forceinline__ int xor_swap(int x, int mask, int dir)
+{
+  int y = __shfl_xor_sync(FULL_MASK, x, mask, WARP_SIZE);
+  return x < y == dir ? y : x;
+}
+
+__device__ __forceinline__ uint bfe(uint lane_id, uint pos)
+{
+  uint res;
+  asm("bfe.u32 %0,%1,%2,%3;" : "=r"(res) : "r"(lane_id), "r"(pos), "r"(1));
+  return res;
+}
+
+// https://en.wikipedia.org/wiki/Xorshift#xorshift*
+__host__ __device__ __forceinline__ uint64_t xorshift64(uint64_t x)
+{
+  x ^= x >> 12;
+  x ^= x << 25;
+  x ^= x >> 27;
+  return x * 0x2545F4914F6CDD1DULL;
+}
+
+template <typename T>
+__device__ __forceinline__ void warp_bitonic_sort(T* element_ptr, const int lane_id)
```

Comment:
In a follow up we shall check if this can be replaced by the our existing bitonic sort.

Reply:
It's a simple version of bitonic sort, it sorts 32 elements with 32 threads.

Reply:
Thanks @RayWang96 for the clarification, we will then bring in RAFT's bitonic sort in a follow up PR.

Reply:
@divyegala please add todo comment for this and reference the GitHub issue in the comment.